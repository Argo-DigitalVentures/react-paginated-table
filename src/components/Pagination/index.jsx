import React, { PropTypes } from 'react';
import range from 'lodash.range';

import PreviousComponent from '../Previous';
import NextComponent from '../Next';

import {
  styleType,
  classNameType,
} from '../../propTypes';

function Pagination(props) {
  const {
    style,
    className,
    count,
    page,
    pageSize,
    showPageJump,
    showPageSizeOptions,
    pageSizeOptions,
    PreviousComponent,
    NextComponent,
    previousStyle,
    nextStyle,
    previousClassName,
    nextClassName,
    setPageSize,
    jumpToPage,
    incrementPage,
    decrementPage,
    ...rest
  } = props;
  const numberOfPages = Math.ceil(count / pageSize);
  return (
    <div style={style} className={className} {...rest}>
      <PreviousComponent
        style={previousStyle}
        className={previousClassName}
        onClick={decrementPage}
        disabled={page <= 1}
      />
      {
        showPageJump ? (
          <div>
            {'Page '}
            <select
              value={page}
              onChange={e => jumpToPage(Number(e.target.value))}
            >
              {
                range(1, numberOfPages + 1).map(option => (
                  <option key={option} value={option}>{option}</option>
                ))
              }
            </select>
            {` of ${Math.ceil(count / pageSize)}`}
          </div>
        ) : null
      }
      <NextComponent
        style={nextStyle}
        className={nextClassName}
        onClick={incrementPage}
        disabled={page >= numberOfPages}
      />
      {
        showPageSizeOptions ? (
          <div>
            <select value={pageSize} onChange={e => setPageSize(Number(e.target.value))}>
              {
                pageSizeOptions.map(pageSize => (
                  <option key={pageSize} value={pageSize}>{pageSize}</option>
                ))
              }
            </select>
          </div>
        ) : null
      }
    </div>
  );
}

Pagination.defaultProps = {
  style: {},
  className: '',
  showPageJump: true,
  showPageSizeOptions: true,
  pageSizeOptions: [5, 10, 20, 25, 50, 100],
  PreviousComponent,
  NextComponent,
  previousStyle: {},
  nextStyle: {},
  previousClassName: '',
  nextClassName: '',
};

Pagination.propTypes = {
  count: PropTypes.number.isRequired,
  page: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  setPageSize: PropTypes.func.isRequired,
  jumpToPage: PropTypes.func.isRequired,
  incrementPage: PropTypes.func.isRequired,
  decrementPage: PropTypes.func.isRequired,
  style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  className: PropTypes.string,
  showPageJump: PropTypes.bool,
  showPageSizeOptions: PropTypes.bool,
  pageSizeOptions: PropTypes.arrayOf(PropTypes.number),
  PreviousComponent: PropTypes.func.isRequired,
  NextComponent: PropTypes.func.isRequired,
  previousStyle: styleType,
  nextStyle: styleType,
  previousClassName: classNameType,
  nextClassName: classNameType,
};

export default Pagination;
