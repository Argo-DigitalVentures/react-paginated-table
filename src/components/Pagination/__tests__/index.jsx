import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Pagination from '../';

describe('<Pagination />', () => {
  it('should accept className, style, and other DOM props', () => {
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should allow the page jump to be hidden', () => {
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
        showPageJump={false}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should allow the page size selector to be hidden', () => {
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
        showPageSizeOptions={false}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should allow custom page size options', () => {
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
        pageSizeOptions={[5, 10, 30, 60]}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should pass custom styles and classNames to Next and Previous components', () => {
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
        previousStyle={{
          color: 'blue',
        }}
        nextStyle={{
          color: 'red',
        }}
        previousClassName="testPrevious"
        nextClassName="testNext"
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should use custom Next and Previous components', () => {
    const CustomPrevious = () => null;
    const CustomNext = () => null;
    const fakeFn = () => null;
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
        PreviousComponent={CustomPrevious}
        NextComponent={CustomNext}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should call the jumpToPage function when new page is selected', () => {
    const fakeFn = () => null;
    const jumpToPageSpy = jest.fn();
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={fakeFn}
        jumpToPage={jumpToPageSpy}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
      />,
    );
    const pageNumberSelect = tree.find('select').at(0);
    pageNumberSelect.simulate('change', {
      target: {
        value: '3',
      },
    });
    expect(jumpToPageSpy).toBeCalledWith(3);
  });

  it('should call the setPageSize function when new page size is selected', () => {
    const fakeFn = () => null;
    const setPageSizeSpy = jest.fn();
    const tree = shallow(
      <Pagination
        style={{
          color: 'green',
        }}
        className="test"
        count={20}
        page={2}
        pageSize={5}
        setPageSize={setPageSizeSpy}
        jumpToPage={fakeFn}
        incrementPage={fakeFn}
        decrementPage={fakeFn}
      />,
    );
    const pageNumberSelect = tree.find('select').at(1);
    pageNumberSelect.simulate('change', {
      target: {
        value: '25',
      },
    });
    expect(setPageSizeSpy).toBeCalledWith(25);
  });
});
