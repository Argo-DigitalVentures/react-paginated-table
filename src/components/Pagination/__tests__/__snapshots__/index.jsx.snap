exports[`<Pagination /> should accept className, style, and other DOM props 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Previous
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    Page 
    <select
      onChange={[Function]}
      value={2}>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
    </select>
     of 4
  </div>
  <Next
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    <select
      onChange={[Function]}
      value={5}>
      <option
        value={5}>
        5
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={20}>
        20
      </option>
      <option
        value={25}>
        25
      </option>
      <option
        value={50}>
        50
      </option>
      <option
        value={100}>
        100
      </option>
    </select>
  </div>
</div>
`;

exports[`<Pagination /> should allow custom page size options 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Previous
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    Page 
    <select
      onChange={[Function]}
      value={2}>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
    </select>
     of 4
  </div>
  <Next
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    <select
      onChange={[Function]}
      value={5}>
      <option
        value={5}>
        5
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={30}>
        30
      </option>
      <option
        value={60}>
        60
      </option>
    </select>
  </div>
</div>
`;

exports[`<Pagination /> should allow the page jump to be hidden 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Previous
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <Next
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    <select
      onChange={[Function]}
      value={5}>
      <option
        value={5}>
        5
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={20}>
        20
      </option>
      <option
        value={25}>
        25
      </option>
      <option
        value={50}>
        50
      </option>
      <option
        value={100}>
        100
      </option>
    </select>
  </div>
</div>
`;

exports[`<Pagination /> should allow the page size selector to be hidden 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Previous
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    Page 
    <select
      onChange={[Function]}
      value={2}>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
    </select>
     of 4
  </div>
  <Next
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
</div>
`;

exports[`<Pagination /> should pass custom styles and classNames to Next and Previous components 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Previous
    className="testPrevious"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "color": "blue",
      }
    } />
  <div>
    Page 
    <select
      onChange={[Function]}
      value={2}>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
    </select>
     of 4
  </div>
  <Next
    className="testNext"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "color": "red",
      }
    } />
  <div>
    <select
      onChange={[Function]}
      value={5}>
      <option
        value={5}>
        5
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={20}>
        20
      </option>
      <option
        value={25}>
        25
      </option>
      <option
        value={50}>
        50
      </option>
      <option
        value={100}>
        100
      </option>
    </select>
  </div>
</div>
`;

exports[`<Pagination /> should use custom Next and Previous components 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <CustomPrevious
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    Page 
    <select
      onChange={[Function]}
      value={2}>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
    </select>
     of 4
  </div>
  <CustomNext
    className=""
    disabled={false}
    onClick={[Function]}
    style={Object {}} />
  <div>
    <select
      onChange={[Function]}
      value={5}>
      <option
        value={5}>
        5
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={20}>
        20
      </option>
      <option
        value={25}>
        25
      </option>
      <option
        value={50}>
        50
      </option>
      <option
        value={100}>
        100
      </option>
    </select>
  </div>
</div>
`;
